### PR TITLE
STOR-5489: Generate retry metadata at actor fetch call sites

### PR DESCRIPTION
**Pull request by apeacock1991**
Generate caller-owned tokens for rewindable global Durable Object fetches. Mark other Fetchers retry-ineligible and reject metadata at unsupported call sites, which will be added later where possible (e.g. RPC will follow in future).

### CODE REVIEW

File: src/workerd/api/actor-state.c++ [hunk 1]
```diff
@@ -242,7 +242,8 @@ DurableObjectStorage::DurableObjectStorage(jsg::Lock& js,
       : Fetcher::RequiresHostAndProtocol::NO;
 
   this->maybePrimary = js.alloc<DurableObject>(
-      js.alloc<DurableObjectId>(kj::mv(primaryActorId)), kj::mv(outgoingFactory), requiresHost);
+      js.alloc<DurableObjectId>(kj::mv(primaryActorId)), kj::mv(outgoingFactory), requiresHost,
+      ActorRetryEligibility::INELIGIBLE);
```
Comment (a-robinson) [inline src/workerd/api/actor-state.c++]: Why shouldn't this be eligible for retries? I don't know why we'd want to exclude replicas' primary stubs.

File: src/workerd/api/container.c++ [hunk 2]
```diff
@@ -1456,8 +1456,8 @@ jsg::Ref<Fetcher> Container::getTcpPort(jsg::Lock& js, int port) {
   kj::Own<Fetcher::OutgoingFactory> factory = kj::heap<TcpPortOutgoingFactory>(
       ioctx.getEntropySource(), ioctx.getHeaderTable(), kj::mv(portState));
 
-  return js.alloc<Fetcher>(
-      ioctx.addObject(kj::mv(factory)), Fetcher::RequiresHostAndProtocol::YES, true);
+  return js.alloc<Fetcher>(ioctx.addObject(kj::mv(factory)),
+      Fetcher::RequiresHostAndProtocol::YES, ActorRetryEligibility::INELIGIBLE, true);
```
Comment (a-robinson) [inline src/workerd/api/container.c++]: Why shouldn't we allow retries when connecting to containers?

File: src/workerd/api/fetch-body-rewindable-test.c++ [hunk 3]
```diff
@@ -222,16 +227,58 @@ KJ_TEST("fetch reports each outgoing body's rewindability per-call without stale
...
+        env.context.addObject<Fetcher::OutgoingFactory>(
+            kj::heap<RetryMetadataOutgoingFactory>(ordinaryDispatchCalled, capturedMetadata)),
+        Fetcher::RequiresHostAndProtocol::YES, ActorRetryEligibility::ELIGIBLE);
+    auto promise = fetcher->fetch(env.js, kj::str("http://example.com"), kj::none);
+    return env.context.awaitJs(env.js, kj::mv(promise)).ignoreResult().attach(kj::mv(fetcher));
+  });
+
+  KJ_EXPECT(!ordinaryDispatchCalled);
+  KJ_IF_SOME(metadata, capturedMetadata) {
+    KJ_EXPECT(metadata.isRetry == IsActorRetry::NO);
```
Comment (a-robinson) [inline src/workerd/api/fetch-body-rewindable-test.c++:261]: Seems worth also checking for a non-empty retry token? I suppose the next test shows that `capturedMetadata` is none when the relevant code isn't triggered, but seeing IsActorRetry::NO isn't all that convincing.

**Review by a-robinson — APPROVED** [review]

File: src/workerd/api/actor-state.c++ (hunk 1 above)
Comment (apeacock1991) [reply src/workerd/api/actor-state.c++]: Ah, I thought this was multi-replica actors but seems it's not. I think I'll need to forward metadata for `ReplicaActorOutgoingFactory` too!

File: src/workerd/api/container.c++ (hunk 2 above)
Comment (apeacock1991) [reply src/workerd/api/container.c++]: I think this is HTTP over TCP via connect(), do we want that included? As I think this targets the container directly? (this is what Sol seems to think, at least)

Comment (a-robinson) [reply src/workerd/api/container.c++]: You're right, sorry. I thought this was getting a DO stub to the container DO, but now see that DOs aren't even involved here at all (other than in the sense that this API is used from within a DO).

This is correct as is 👍

File: src/workerd/api/fetch-body-rewindable-test.c++ (hunk 3 above)
Comment (apeacock1991) [reply src/workerd/api/fetch-body-rewindable-test.c++:261]: Yup that's fair, fixed!

File: src/workerd/api/actor-state.c++ (hunk 1 above)
Comment (apeacock1991) [reply src/workerd/api/actor-state.c++]: This is now included!

**Review by a-robinson — APPROVED** [review]
👍
